ably.ts.d

```diff
@@ -0,0 +1,400 @@
+  on: (eventOrCallback: String | ListenerFn, callback?: ListenerFn) => void;
+
+  once: (eventOrCallback: String | ListenerFn, callback?: ListenerFn) => void;
+
+  off: (eventOrCallback?: String | ListenerFn, callback?: ListenerFn) => void;
+}
+
+// Classes
+export declare class Auth {
+  clientId: String;
+  authorize: (tokenParams?: TokenParams, authOptions?: AuthOptions, callback?: (Error: ErrorInfo, Results: TokenDetails) => void) => void;
+  createTokenRequest: (tokenParams?: TokenParams, authOptions?: AuthOptions, callback?: (Error: ErrorInfo, Results: TokenRequest) => void) => void;
+  requestToken: (TokenParams?: TokenParams, authOptions?: AuthOptions, callback?: (Error: ErrorInfo, Results: TokenDetails) => void) => void;
+}
+
+export declare class Presence {
+  get: (params: RestPresenceParams, PaginatedResultCallback) => void;
+
+  history: (params: RestPresenceHistoryParams, PaginatedResultCallback) => void;
+}
+
+export declare class RealtimePresence {
+  syncComplete: () =>  boolean;
+
+  get: (Params: RealtimePresenceParams, callback?: PaginatedResultCallback) => void;
+
+  history: (ParamsOrCallback: RealtimePresenceHistoryParams | PaginatedResultCallback, callback?: PaginatedResultCallback) => void;
+
+  subscribe: (presenceOrCallback: PresenceAction | messageCallback, listener?: standardCallback, callback?: standardCallback) => void;
+
+  unsubscribe: (presence?: PresenceAction, listener?: standardCallback, callback?: standardCallback) => void;
+
+  enter: (Data: any, callback?: standardCallback) => void;
+
+  update: (Data: any, callback?: standardCallback) => void;
+
+  leave: (Data: any, callback?: standardCallback) => void;
+
+  enterClient: (clientId: String, Data: any, callback?: standardCallback) => void;
+
+  updateClient: (clientId: String, Data: any, callback?: standardCallback) => void;
+
+  leaveClient: (clientId: String, Data: any, callback?: standardCallback) => void;
+}
+
+export declare class Channel {
+  name: String;
+  presence: Presence;
+  history: (ParamsOrCallback?: RestPresenceHistoryParams | PaginatedResultCallback, callback?: PaginatedResultCallback ) => void;
+  publish: (messages: any, MessageDataOrCallback?: any | errorCallback, callback?: errorCallback) => void;
+}
+
+export declare class RealtimeChannel extends EventEmitter {
+  name: String;
+  errorReason: ErrorInfo;
+  state: ChannelState;
+  presence: RealtimePresence;
+
+  attach: (callback?: standardCallback) => void;
+
+  detach:(callback?: standardCallback) => void;
+
+  history: (ParamsOrCallback?: RealtimePresenceHistoryParams | PaginatedResultCallback, callback?: PaginatedResultCallback) => void;
+
+  subscribe: (eventOrCallback: String | messageCallback, listener?: messageCallback) => void;
+
+  unsubscribe: (eventOrCallback?: String | messageCallback, listener?: messageCallback) => void;
+
+  publish: (messages: any, MessageDataOrCallback?: any | errorCallback, callback?: errorCallback) => void;
+
+}
+
+export declare class Channels {
```

Comment:
This needs to support generics so that `Channels` can return `RealtimeChannel` or `Channel` objects